src/som/interpreter/processes/SChannel.java

```diff
@@ -70,27 +76,55 @@ public static SChannelInput create(final SynchronousQueue<Object> cell,
 
     private final SynchronousQueue<Object> cell;
     protected final SChannel               channel;
+    private final AtomicInteger            numReads;
 
     public SChannelInput(final SynchronousQueue<Object> cell,
         final SChannel channel) {
       this.cell = cell;
       this.channel = channel;
+      if (VmSettings.ACTOR_TRACING | VmSettings.REPLAY) {
+        numReads = new AtomicInteger(0);
+      } else {
+        numReads = null;
+      }
     }
 
     @TruffleBoundary
-    public Object read() throws InterruptedException {
+    public Object read(final RecordTwoEvent traceRead) throws InterruptedException {
       ObjectTransitionSafepoint.INSTANCE.unregister();
+
+      if (VmSettings.REPLAY) {
+        Activity reader = TracingActivityThread.currentThread().getActivity();
+        ChannelReadRecord crr = (ChannelReadRecord) reader.getNextReplayEvent();
+        assert crr != null : reader;
+        assert channel.getId() == crr.channelId;
+
+        while (numReads.get() != crr.readNo) {
```

Comment:
can you factor this out into a `replayDelay(currentId, expectedId, currentChannelId, expectedChannelId)` or so method? and use it consistently for all the places with such asserts/loops?

I hope this would make the code more readable, and avoids undesirable inconsistencies.

Reply:
Made it event more generic than suggested, can be reused for any numbered events in passive entities.